connectors/storages.mongodb/src/storage.js

```diff
@@ -45,28 +54,44 @@ class Storage extends Connector {
   }
 
   async set (entity) {
-    const criteria = { _id: entity.id, _version: entity._version }
+    const criteria = {
+      _id: entity.id,
+      _version: entity._version - 1
+    }
     const result = await this.#connection.replace(criteria, to(entity))
 
     return result !== null
   }
 
   async store (entity) {
-    if (entity._version === 0) return this.add(entity)
-    else return this.set(entity)
+    if (entity._version === 1) {
+      return this.add(entity)
+    } else {
+      return this.set(entity)
+    }
   }
 
   async upsert (query, changeset, insert) {
-    const { criteria, options } = translate(query)
-    const update = { $set: { ...changeset }, $inc: { _version: 1 } }
+    const {
+      criteria,
+      options
+    } = translate(query)
+
+    const update = {
+      $set: { ...changeset },
+      $inc: { _version: 1 }
+    }
 
     if (insert !== undefined) {
       delete insert._version
 
       options.upsert = true
 
-      if (criteria._id !== undefined) insert._id = criteria._id
-      else return null // this shouldn't ever happen
+      if (criteria._id !== undefined) {
+        insert._id = criteria._id
+      } else {
+        return null
+      } // this shouldn't ever happen
```

Comment:
why not `assert` then?

```js
assert.notStrictEqual(criteria._id, undefined, 'This should never happen')
insert._id = criteria._id
```

Reply:
I don't want to touch this code :)
It will be eventually rewritten to TS.